pkg/scheduler/controller/shoot/reconciler.go

```diff
@@ -318,6 +326,48 @@ func filterSeedsWithDisabledShootReconciliations(seedList []gardencorev1beta1.Se
 	return seedsWithEnabledReconciliations, nil
 }
 
+// filterSeedsMatchingDomain filters seeds that can support the shoot's domain configuration.
+// If the shoot uses a default domain, only seeds that have that domain configured in their DNS defaults are selected.
+// If the shoot uses a custom domain, all seeds are accepted.
+func filterSeedsMatchingDomain(seedList []gardencorev1beta1.Seed, shoot *gardencorev1beta1.Shoot, projectName string) ([]gardencorev1beta1.Seed, error) {
+	// If the shoot doesn't have a domain specified, all seeds are eligible
+	if shoot.Spec.DNS == nil || shoot.Spec.DNS.Domain == nil {
+		return seedList, nil
+	}
+
+	var (
+		shootDomain              = *shoot.Spec.DNS.Domain
+		supportingSeeds          []gardencorev1beta1.Seed
+		hasMatchedGardenerDomain bool
+	)
+
+	for _, seed := range seedList {
+		if len(seed.Spec.DNS.Defaults) > 0 {
+			for _, defaultDomain := range seed.Spec.DNS.Defaults {
```

Comment:
You can skip the `if` check, the for loop will not be executed if the Defaults slice is empty. 
```suggestion
		for _, defaultDomain := range seed.Spec.DNS.Defaults {
```